ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/attr/Attr.java

```diff
@@ -0,0 +1,154 @@
+  static class Impl<T> implements TableUdaf<T, List<Struct>, T> {
+
+    static final String VALUE = "VALUE";
+    static final String COUNT = "COUNT";
+
+    SqlType inType;
+    Schema entrySchema;
+
+    @Override
+    public void initializeTypeArguments(final List<SqlArgument> args) {
+      this.inType = args.get(0).getSqlTypeOrThrow();
+
+      // we use a list of structs instead of a map here for two reasons:
+      //
+      //   1. our data formats currently only support maps with string keys
+      //   2. null is a valid entry, and most maps don't support null keys
+      //
+      // this should be OK from a complexity perspective because ATTR expects
+      // a unique entry, so only the situations where it is used improperly will
+      // take a runtime hit
+      this.entrySchema = SchemaBuilder.struct()
+          .optional()
+          .field(VALUE, SchemaConverters.sqlToConnectConverter().toConnectSchema(inType))
+          .field(COUNT, Schema.OPTIONAL_INT32_SCHEMA)
+          .build();
+    }
+
+    @Override
+    public Optional<SqlType> getAggregateSqlType() {
+      return Optional.of(SqlTypes.array(
+          SchemaConverters.connectToSqlConverter().toSqlType(entrySchema)
+      ));
+    }
+
+    @Override
+    public Optional<SqlType> getReturnSqlType() {
+      return Optional.of(inType);
+    }
+
+    @Override
+    public List<Struct> initialize() {
+      return new ArrayList<>();
+    }
+
+    @Override
+    public List<Struct> aggregate(final T current, final List<Struct> agg) {
+      final List<Struct> out = new ArrayList<>(agg);
+      update(out, current, 1);
+      return out;
+    }
+
+    @Override
+    public List<Struct> merge(final List<Struct> one, final List<Struct> two) {
+      // use O(n^2) algorithm here because in practice each of these lists
+      // should have no more than one entry (otherwise it's an invalid Attr
+      // anyway)
+      final List<Struct> out = new ArrayList<>(one);
+      for (final Struct entry : two) {
+        update(out, entry.get(VALUE), entry.getInt32(COUNT));
+      }
+      return out;
+    }
+
+    @Override
+    public List<Struct> undo(final T valueToUndo, final List<Struct> agg) {
+      final List<Struct> out = new ArrayList<>(agg);
+      update(out, valueToUndo, -1);
+      return out;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T map(final List<Struct> agg) {
+      final List<Struct> collect = agg.stream()
+          .filter(s -> s.getInt32(COUNT) > 0)
+          .collect(Collectors.toList());
+
+      if (collect.size() != 1) {
+        return null;
+      }
+
+      return (T) collect.get(0).get(VALUE);
+    }
+
+    private void update(final List<Struct> agg, final Object current, final int count) {
+      boolean found = false;
+      for (final Struct entry : agg) {
+        if (Objects.equals(entry.get(VALUE), current)) {
+          found = true;
+          entry.put(COUNT, Math.max(0, entry.getInt32(COUNT) + count));
```

Comment:
Can we `break;` and stop iterating once we've found the element?

If there can be multiple copies in the list, then this would update all of them by 1!

Reply:
hah! you're right 😉 don't think it should ever come up in practice but I'll go ahead and update this

Reply:
ah man... if only I had responded to this with "can't catch a break, can I?"

Reply:
Next time, if you try exceptionally hard, you'll find the resources to make bad puns more readily before the close of business....